towhee/dataframe/dataframe_v2.py

```diff
@@ -56,23 +57,51 @@ def __init__(
         self._data_as_list = None
         self._data_as_dict = None
 
-        # TODO: Enforce columns everytime except for when dict passed in.
-        # or supply default 'col_x' naming convention.
-        # TODO: This is equivalent to @junjie's schema
-        if columns is not None:
-            self._types = {x[0]: x[1] for x in columns}
-            self._columns = [x[0] for x in columns]
 
-        elif not isinstance(data, dict):
-            raise ValueError(
-                    'Cannot construct dataframe without colum names and types (except for dict).')
+        self._schema = _Schema()
 
-        # For `data` is empty
-        if not data:
+        # TODO: Better solution for no data whatsoever
+        # if no columns and no data: delay to first data added
+        if columns is None and data is None:
+            raise ValueError('Cannot construct dataframe without columns or initial data')
+
+        # if column and no data: create empty arrays
+        elif columns is not None and data is None:
+            self._set_cols(columns)
             self._from_none()
 
+        # if no column and data: default names and types from data
+        elif columns is None and data is not None:
+            self._extract_data(data, default_cols=True)
+            self._schema_update_needed = True
+            self._update_schema()
+
+        # if column and data: create regular running.
+        elif columns is not None and data is not None:
+            self._set_cols(columns)
+            self._extract_data(data, default_cols=False)
+
+
+    def _from_none(self):
+        self._data_as_list = [Array(name=name) for name, _ in self._schema.cols]
+        self._data_as_dict = {name: self._data_as_list[i] for i, (name, _) in enumerate(self._schema.cols)}
+
+    def _set_cols(self, columns):
+        for names, types in columns:
+                self._schema.add_col(name=names, col_type = types)
+
+    def _update_schema(self):
```

Comment:
We wouldn't change schema when putting data. The schema will not be changed after dataframe `__init__`.